SPEC.md

```diff
@@ -0,0 +1,28 @@
+# Navigation.js specification
+
+## Purpose
+The goal of this specification is to outline the rules and ideas this library adheres to guide a user along a route.
+
+### Units
+By default, the metric system should be used where available.
+
+### Concepts
+Navigation.js looks at a single step along a greater route. All calculations are done against this step and not against the whole route. Once a step is completed, the next step on the route can be compared. If the user is no longer on the current step, the developer should re-route the user and also set their internal step counter back to 0. The developer is required to keep track of the current step at all times.
+
+### Re-routing
+If at any time the user is not within `45 meters` of any point on the current step, re-route the user.
+
```

Comment:
For reference, the standard width of a U.S. Interstate lane is 12 feet (3.7 meters), while surface streets typically have narrower lanes. A typical intersection between two surface streets that each have five lanes (two in either direction, plus a left turn lane in the middle) might be about 102–120 feet (31.1–36.6 meters) across, [depending on skew](http://ops.fhwa.dot.gov/publications/fhwahop08024/chapter3.htm) (see fig. 3-1).

Considering that the Mapbox Directions API curently returns a linear route line with no lane counts or lane widths, a threshold of 45 meters allows for the driver to be 12 lanes away from the route line in either direction, or anywhere within an intersection between streets that have 24 lanes. @MoKob is right that this threshold can be tightened somewhat.

As for wrong turns, we should immediately reroute when the user begins to turn in a direction opposite what we expect for a given maneuver, subject to a threshold that allows for GPS uncertainty. I think that should be part of whatever heuristics we have in the “Completing a step” section, but the “Re-routing” component is primarily about deviating from the route partway through a step.